compiler/vm/vm.nim

```diff
@@ -267,71 +432,104 @@ proc copyValue(src: PNode): PNode =
     for i in 0..<src.len:
       result[i] = copyValue(src[i])
 
-proc asgnComplex(x: var TFullReg, y: TFullReg) =
-  x.ensureKind(y.kind)
-  case x.kind
+proc asgnComplex(mm: var VmMemoryManager, x: var TFullReg, y: TFullReg) =
+  ## If register `y` is set, transition `x` into a value register and
+  ## perform a deep copy of `y`'s data
+  case y.kind
   of rkNone: discard
-  of rkInt: x.intVal = y.intVal
-  of rkFloat: x.floatVal = y.floatVal
-  of rkNode: x.node = copyValue(y.node)
-  of rkRegisterAddr: x.regAddr = y.regAddr
-  of rkNodeAddr: x.nodeAddr = y.nodeAddr
+  of rkInt: x = TFullReg(kind: rkInt, intVal: y.intVal)
+  of rkFloat: x = TFullReg(kind: rkFloat, floatVal: y.floatVal)
+  of rkAddress: x = TFullReg(kind: rkAddress, addrVal: y.addrVal, addrTyp: y.addrTyp, isValid: y.isValid)
```

Comment:
These can be condensed down to:

```nim
x = y
````

As objects are copied by default.

Specialize only when you need to adjust the resulting value, like what you do with `rkHandle`/`rkValue` for example.

Reply:
In the code you're looking at, no. Copying is unavailable for `TFullReg` due to the copy hook being disabled with `{.error.}`. This is mostly meant as way to force code authors to do the copy manually in order to ensure that it's really what is wanted and to prevent accidental copies which could result in bugs.

Unfortunately, this triggered a code-gen issue in the csources compiler which resulted in a very-hard-to-track-down memory corruption issue (`x` is not reset prior to moving into it), so I'm going to remove the restriction (otherwise, I'd have to disable the `=sink` hook which would then require manual field assignment), which will make `x = y` possible